etc/ign.bash_completion.sh

```diff
@@ -1,31 +1,51 @@
+#!/usr/bin/env bash
+
 # ign bash completion
 
 function _ign
 {
   local prev cur cmd opts
   local ign="$1"
-  COMPREPLY=()
   cur="$2"
   prev="$3"
 
-  # searching for the command
+  # Return value
+  COMPREPLY=()
+
+  # Look for the first command that is not an option (-*)
+  # COMP_WORDS: array of words in current command line COMP_LINE
+  # COMP_CWORD: index of word containing current cursur location
+  # COMP_LINE: line entered so far
   for ((i=1; $i<=$COMP_CWORD; i++)); do
     if [[ ${COMP_WORDS[i]} != -* ]]; then
       cmd="${COMP_WORDS[i]}"
       break
     fi
   done
 
+  # On a word after top-level command ign. It may be an option (-*)
   if [[ "$cur" == -* ]] || [[ "$prev" != "ign" ]]; then
 
+    # Subcommand is help
     if [[ "$cmd" == "help" ]]; then
       opts=$(ign --commands)
+
+    # Subcommand is a library name or an option (-*)
     else
-      COMPREPLY=($(compgen -f  -- "${COMP_WORDS[${COMP_CWORD}]}" ))
-      complete -o filenames -o nospace -F "_ign" "ign"
-      return
+      # Complete subcommands
+      case "$cmd" in
+        gui)
+          _gz_gui
+          return
+          ;;
+        plugin)
+          _gz_plugin
```

Comment:
Ideally we wouldn't hardcode the names of downstream libraries here, this creates a circular dependency. This also breaks if we source some libraries but not others. For example, I only sourced the script from plugin, and tried this:

```
$ ign gui --bash: _gz_gui: command not found
```

Something like this may work, the idea is to use `$cmd` to generate the function name, and then only call it if it's defined

```
      func="_gz_$cmd"
      if [[ "$(type -t $func)" == 'function' ]]; then 
        $func
      fi
```

Reply:
Thanks for the snippet. That works! 847713d